engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/UniqueItemsValidator.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Hibernate Validator, declare and validate application constraints
+ *
+ * License: Apache License, Version 2.0
+ * See the license.txt file in the root directory or <http://www.apache.org/licenses/LICENSE-2.0>.
+ */
+package org.hibernate.validator.internal.constraintvalidators.hv;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.validation.ConstraintValidator;
+import javax.validation.ConstraintValidatorContext;
+
+import org.hibernate.validator.constraints.UniqueItems;
+
+/**
+ * Validates that the provided collection is a set of unique items
+ * <p>
+ * Uniqueness is defined by the equals method of the item being compared.
+ *
+ * @author Tadhg Pearson
+ */
+public class UniqueItemsValidator implements ConstraintValidator<UniqueItems, Collection<?>> {
+	@Override
+	public void initialize(UniqueItems uniqueItems) {
+		//No annotation fields to initialize
+	}
+
+	/**
+	 * @param objects Potentially null collection of input objects to check for duplicates
+	 * @param validatorContext Non-null {@link ConstraintValidatorContext} to accumulate validation messages
+	 *
+	 * @return True if the input collection is null or does not contain duplicate objects
+	 */
+	@Override
+	public boolean isValid(Collection<?> objects, ConstraintValidatorContext validatorContext) {
+		boolean valid = isValid( objects );
+		if ( !valid ) {
+			List<String> duplicates = findDuplicates( objects );
+			String message = "Contains the following duplicate items: " + duplicates;
+			validatorContext.disableDefaultConstraintViolation();
+			validatorContext.buildConstraintViolationWithTemplate( message ).addConstraintViolation();
+		}
+		return valid;
+	}
+
+	private boolean isValid(Collection<?> objects) {
+		boolean valid;
+		if ( objects == null ) {
+			valid = true;
+		}
+		else {
+			Set<?> set = new HashSet<>( objects );
+			valid = set.size() == objects.size();
+		}
+		return valid;
+	}
+
+	/**
+	 * Find the duplicate items the provided input collection. The duplicate items will be presented in the order in which
+	 * they occur. Duplicate items which occur multiple times will be presented multiple times in the duplicates list.
+	 *
+	 * @param objects Non-null collection of objects in which to find duplicates.
+	 *
+	 * @return Non-null collection of the toString output for the dupliciate objects.
+	 */
```

Comment:
I don't think we need a comment for such a trivial private method. Let's keep it simple :).